incubator/hnc/pkg/controllers/object_controller_test.go

```diff
@@ -120,6 +120,65 @@ var _ = Describe("Secret", func() {
 		Eventually(isModified(ctx, barName, "foo-sec")).Should(BeTrue())
 		Eventually(isModified(ctx, bazName, "foo-sec")).Should(BeTrue())
 	})
+
+	It("shouldn't propagate/delete if the namespace has Crit condition", func() {
+		if !newObjectController {
+			return
+		}
+
+		// Set tree as baz -> bar -> foo(root).
+		setParent(ctx, barName, fooName)
+		setParent(ctx, bazName, barName)
+
+		Eventually(hasSecret(ctx, barName, "foo-sec")).Should(BeTrue())
+		Expect(secretInheritedFrom(ctx, barName, "foo-sec")).Should(Equal(fooName))
+
+		Eventually(hasSecret(ctx, bazName, "foo-sec")).Should(BeTrue())
+		Expect(secretInheritedFrom(ctx, bazName, "foo-sec")).Should(Equal(fooName))
+
+		Eventually(hasSecret(ctx, bazName, "bar-sec")).Should(BeTrue())
+		Expect(secretInheritedFrom(ctx, bazName, "bar-sec")).Should(Equal(barName))
+
+		// Set parent to a non-existent namespace.
+		brumpfName := createNSName("brumpf")
+		barHier := newOrGetHierarchy(ctx, barName)
+		barHier.Spec.Parent = brumpfName
+		updateHierarchy(ctx, barHier)
+		Eventually(hasCondition(ctx, barName, api.CritParentMissing)).Should(Equal(true))
+		Eventually(hasCondition(ctx, bazName, api.CritAncestor)).Should(Equal(true))
+
+		// Everything should remain the same in the namespaces with Crit condition.
```

Comment:
I wonder if we can make this a bit more compelling and also more realistic for a controller restart case.

We start with baz -> bar -> foo, with baz inheriting a secret from both bar and foo. Then set the parent of foo to brumph; nothing should change. But then again, even if we _were_ syncing, nothing would change anyway, so this isn't a very strong test.

Now set the parent of baz to foo. If we were syncing, baz would lose its secret from bar, but because syncing is frozen, it should still stick around. We could also maybe add a second secret to foo and show that it's not propagated either. (It's hard to prove that something _doesn't_ happen. Maybe wait one second before an Expect() clause. And then manually comment out your new code and show that the test now fails).

Then, finally, create the brumpth namespace and make sure everything gets resolved.

Make sense?

Reply:
Yes, I see. Just confirming, your point is to make a real missing parent (foo -> "brumph") just like in controller restart phase instead of changing current parent to a missing namespace (from baz -> foo to baz -> "brumph")?

The whole test sgtm and I will go ahead to change it. And adding a new foo object sg. It's more explicit and therefore, we test both no-deleting and no-propagating.

Reply:
yup sgtm